// decentraland/dist/commands/deploy.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const wrapCommand_1 = require("../utils/wrapCommand");
const logging_1 = require("../utils/logging");
const analytics_1 = require("../utils/analytics");
const Decentraland_1 = require("../lib/Decentraland");
const opn = require("opn");
const inquirer = require("inquirer");
const errors_1 = require("../utils/errors");
const MAX_FILE_COUNT = 100;
function deploy(vorpal) {
    vorpal
        .command('deploy')
        .description('Uploads scene to IPFS and updates IPNS.')
        .option('-h, --host <string>', 'IPFS daemon API host (default is localhost).')
        .option('-p, --port <number>', 'IPFS daemon API port (default is 5001).')
        .option('-s, --skip', 'skip confirmations and proceed to upload')
        .option('-hs, --https', 'Use self-signed localhost certificate to use HTTPs at linking app (required for ledger users)')
        .action(wrapCommand_1.wrapCommand((args) => __awaiter(this, void 0, void 0, function* () {
        const dcl = new Decentraland_1.Decentraland({
            ipfsHost: args.options.host || 'localhost',
            ipfsPort: args.options.port || 5001,
            isHttps: !!args.options.https
        });
        let ignoreFile = yield dcl.project.getDCLIgnore();
        dcl.on('ipfs:add', () => {
            const spinner = logging_1.loading('Uploading files to local IPFS node');
            dcl.on('ipfs:add-success', () => {
                spinner.succeed();
            });
        });
        dcl.on('ethereum:get-ipns', (x, y) => {
            const spinner = logging_1.loading(`Checking IPNS for coordinates ${x},${y}`);
            dcl.on('ethereum:get-ipns-empty', () => {
                spinner.info(`No IPNS found for coordinates ${x},${y}`);
            });
            dcl.on('ethereum:get-ipns-success', () => {
                spinner.succeed();
            });
        });
        dcl.on('ipfs:publish', (ipfsHash) => {
            const spinner = logging_1.loading(`Publishing IPNS for ${ipfsHash}`);
            dcl.on('ipfs:publish-success', (ipnsHash) => {
                spinner.succeed();
                logging_1.info(`IPNS hash: ${ipnsHash}`);
            });
        });
        dcl.on('link:ready', url => {
            analytics_1.Analytics.sceneLink();
            logging_1.info('This is the first time you deploy using this IPNS, please link your project to the LAND Registry:');
            const linkerMsg = logging_1.loading(`Linking app ready at ${url}`);
            setTimeout(() => {
                try {
                    opn(url);
                }
                catch (e) {
                    vorpal.log(logging_1.warning(`WARNING: Unable to open browser automatically`));
                }
            }, 5000);
            dcl.on('link:success', () => {
                analytics_1.Analytics.sceneLinkSuccess();
                linkerMsg.succeed('Project successfully updated in LAND Registry');
            });
        });
        dcl.on('ipfs:pin', () => {
            analytics_1.Analytics.pinRequest();
            const spinner = logging_1.loading(`Pinning files to IPFS gateway`);
            dcl.on('ipfs:pin-success', () => __awaiter(this, void 0, void 0, function* () {
                analytics_1.Analytics.pinSuccess();
                spinner.succeed();
            }));
        });
        if (args.options.https) {
            vorpal.log(logging_1.warning(`WARNING: Using self signed certificate to support ledger wallet`));
        }
        if (ignoreFile === null) {
            vorpal.log(logging_1.warning(`WARNING: As of version 1.1.0 all deployments require a .dclignore file`));
            logging_1.info(`Generating .dclignore file with default values`);
            ignoreFile = yield dcl.project.writeDclIgnore();
        }
        analytics_1.Analytics.sceneDeploy();
        yield dcl.project.validateExistingProject();
        const files = yield dcl.project.getFiles(ignoreFile);
        vorpal.log('\n  Tracked files:\n');
        const totalSize = files.reduce((size, file) => {
            vorpal.log(`    ${file.path} (${file.size} bytes)`);
            return size + file.size;
        }, 0);
        if (files.length > MAX_FILE_COUNT) {
            errors_1.fail(errors_1.ErrorType.DEPLOY_ERROR, `You cannot upload more than ${MAX_FILE_COUNT} files per scene.`);
        }
        vorpal.log(''); // new line to keep things clean
        if (!args.options.skip) {
            const results = yield inquirer.prompt({
                type: 'confirm',
                name: 'continue',
                default: true,
                message: `You are about to upload ${files.length} files (${totalSize} bytes). Do you want to continue?`
            });
            if (!results.continue) {
                vorpal.log('Aborting...');
                process.exit(1);
            }
        }
        yield dcl.deploy(files);
        analytics_1.Analytics.sceneDeploySuccess();
        vorpal.log(logging_1.positive(`\nDeployment complete!`));
    })));
}
exports.deploy = deploy;
