// eth-connect/dist/utils/config.js
"use strict";
/*
    This file is part of web3.js.

    web3.js is free software: you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    web3.js is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with web3.js.  If not, see <http://www.gnu.org/licenses/>.
*/
Object.defineProperty(exports, "__esModule", { value: true });
/// required to define ETH_BIGNUMBER_ROUNDING_MODE
const bignumber_js_1 = require("bignumber.js");
exports.ETH_UNITS = [
    'wei',
    'kwei',
    'Mwei',
    'Gwei',
    'szabo',
    'finney',
    'femtoether',
    'picoether',
    'nanoether',
    'microether',
    'milliether',
    'nano',
    'micro',
    'milli',
    'ether',
    'grand',
    'Mether',
    'Gether',
    'Tether',
    'Pether',
    'Eether',
    'Zether',
    'Yether',
    'Nether',
    'Dether',
    'Vether',
    'Uether'
];
exports.ETH_PADDING = 32;
exports.ETH_SIGNATURE_LENGTH = 4;
exports.ETH_BIGNUMBER_ROUNDING_MODE = {
    ROUNDING_MODE: bignumber_js_1.BigNumber.ROUND_DOWN
};
exports.ETH_POLLING_TIMEOUT = 1000 / 2;
exports.defaultBlock = 'latest';
