// decentraland/dist/lib/Project.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs = require("fs-extra");
const uuid = require("uuid");
const dockerNames = require("docker-names");
const path = require("path");
const filesystem_1 = require("../utils/filesystem");
const project_1 = require("../utils/project");
const ignore = require("ignore");
const errors_1 = require("../utils/errors");
const coordinateHelpers_1 = require("../utils/coordinateHelpers");
var BoilerplateType;
(function (BoilerplateType) {
    BoilerplateType["TYPESCRIPT_STATIC"] = "ts-static";
    BoilerplateType["TYPESCRIPT_DYNAMIC"] = "ts-dynamic";
    BoilerplateType["WEBSOCKETS"] = "multiplayer";
    BoilerplateType["STATIC"] = "static";
})(BoilerplateType = exports.BoilerplateType || (exports.BoilerplateType = {}));
class Project {
    constructor(workingDir) {
        this.workingDir = workingDir;
    }
    /**
     * Returns `true` if the provided path contains a scene file
     */
    sceneFileExists() {
        return fs.pathExists(project_1.getSceneFilePath(this.workingDir));
    }
    /**
     * Returns `true` if the provided path contains a `.decentraland` folder
     */
    decentralandFolderExists() {
        return fs.pathExists(project_1.getDecentralandFolderPath(this.workingDir));
    }
    /**
     * Returns `true` if the project working directory is empty of files
     */
    isProjectDirEmpty() {
        return __awaiter(this, void 0, void 0, function* () {
            return filesystem_1.isEmptyDirectory(this.workingDir);
        });
    }
    /**
     * Returns `true` for valid boilerplate types (`static`, `ts` and `ws`)
     * @param type
     */
    isValidBoilerplateType(type) {
        for (let key in BoilerplateType) {
            if (type === BoilerplateType[key]) {
                return true;
            }
        }
        return false;
    }
    /**
     * Returns an object containing the contents of the `project.json` file.
     */
    getProjectFile() {
        return __awaiter(this, void 0, void 0, function* () {
            return filesystem_1.readJSON(project_1.getProjectFilePath(this.workingDir));
        });
    }
    /**
     * Returns an object containing the contents of the `scene.json` file.
     */
    getSceneFile() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.sceneFile) {
                return this.sceneFile;
            }
            this.sceneFile = yield filesystem_1.readJSON(project_1.getSceneFilePath(this.workingDir));
            return this.sceneFile;
        });
    }
    /**
     * Creates the `project.json` file and all other mandatory folders.
     * @param dirName The directory where the project file will be created.
     */
    initProject() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.writeProjectFile({ id: uuid.v4(), ipfsKey: null });
        });
    }
    /**
     * Scaffolds a project or fails for an invalid boilerplate type
     * @param boilerplateType `static`, `singleplayer` or `multiplayer`
     * @param websocketServer Optional websocket server URL
     */
    scaffoldProject(boilerplateType, websocketServer) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.isValidBoilerplateType(boilerplateType)) {
                errors_1.fail(errors_1.ErrorType.PROJECT_ERROR, `Invalid boilerplate type: '${boilerplateType}'. Supported types are 'static', 'singleplayer' and 'multiplayer'.`);
            }
            switch (boilerplateType) {
                case BoilerplateType.TYPESCRIPT_STATIC: {
                    yield this.copySample('ts-static');
                    break;
                }
                case BoilerplateType.TYPESCRIPT_DYNAMIC: {
                    yield this.copySample('ts-dynamic');
                    break;
                }
                case BoilerplateType.WEBSOCKETS:
                    yield this.scaffoldWebsockets(websocketServer);
                    break;
                case BoilerplateType.STATIC:
                default:
                    yield this.copySample('basic-static');
                    break;
            }
        });
    }
    /**
     * Returns true if the project contains a package.json file and an empty node_modules folder
     */
    needsDependencies() {
        return __awaiter(this, void 0, void 0, function* () {
            const files = yield this.getAllFilePaths();
            const hasPackageFile = files.some(file => file === 'package.json');
            const nodeModulesPath = path.resolve(this.workingDir, 'node_modules');
            const hasNodeModulesFolder = yield fs.pathExists(nodeModulesPath);
            const isNodeModulesEmpty = (yield this.getAllFilePaths(nodeModulesPath)).length === 0;
            if (hasPackageFile && (!hasNodeModulesFolder || isNodeModulesEmpty)) {
                return true;
            }
            return false;
        });
    }
    /**
     * Returns true if te project root contains a `tsconfig.json` file
     * @param dir
     */
    isTypescriptProject() {
        return __awaiter(this, void 0, void 0, function* () {
            const files = yield this.getAllFilePaths();
            return files.some(file => file === 'tsconfig.json');
        });
    }
    /**
     * Writes the provided websocket server to the `scene.json` file
     * @param server The url to a websocket server
     */
    scaffoldWebsockets(server) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.copySample('websockets');
            if (server) {
                yield this.writeSceneFile({ main: server });
            }
        });
    }
    /**
     * Creates a new `project.json` file
     * @param content The content of the `project.json` file
     */
    writeProjectFile(content) {
        return filesystem_1.writeJSON(project_1.getProjectFilePath(this.workingDir), content);
    }
    /**
     * Creates a new `scene.json` file
     * @param path The path to the directory where the file will be written.
     */
    writeSceneFile(content) {
        return filesystem_1.writeJSON(project_1.getSceneFilePath(this.workingDir), content);
    }
    /**
     * Copies the contents of a specific sample into the project (for scaffolding purposes).
     * Merges `scene.json` and `package.json` files
     * @param project The name of the sample folder (used as an indentifier).
     * @param destination The path to the project root. By default the current woxsrking directory.
     */
    copySample(project) {
        return __awaiter(this, void 0, void 0, function* () {
            const src = path.resolve(__dirname, '..', 'samples', project);
            const files = yield fs.readdir(src);
            for (let i = 0; i < files.length; i++) {
                const file = files[i];
                if (file === project_1.SCENE_FILE) {
                    const sceneFile = yield filesystem_1.readJSON(project_1.getSceneFilePath(src));
                    yield this.writeSceneFile(sceneFile);
                }
                else if (file === project_1.PACKAGE_FILE) {
                    const pkgFile = yield filesystem_1.readJSON(project_1.getPackageFilePath(src));
                    yield filesystem_1.writeJSON(project_1.getPackageFilePath(this.workingDir), pkgFile);
                }
                else {
                    yield fs.copy(path.join(src, file), path.join(this.workingDir, file));
                }
            }
        });
    }
    /**
     * Returns a promise of an object containing the base X and Y coordinates for a parcel.
     */
    getParcelCoordinates() {
        return __awaiter(this, void 0, void 0, function* () {
            const sceneFile = yield this.getSceneFile();
            const { base } = sceneFile.scene;
            return coordinateHelpers_1.getObject(base);
        });
    }
    /**
     * Returns a promise of an array of the parcels of the scene
     */
    getParcels() {
        return __awaiter(this, void 0, void 0, function* () {
            const sceneFile = yield this.getSceneFile();
            return sceneFile.scene.parcels.map(coordinateHelpers_1.getObject);
        });
    }
    /**
     * Returns a promise of the owner address
     */
    getOwner() {
        return __awaiter(this, void 0, void 0, function* () {
            const { owner } = yield this.getSceneFile();
            if (!owner) {
                errors_1.fail(errors_1.ErrorType.PROJECT_ERROR, `Missing owner attribute at scene.json. Owner attribute is required for deploying`);
            }
            return owner.toLowerCase();
        });
    }
    /**
     * Fails the execution if one of the parcel data is invalid
     */
    validateParcelOptions() {
        return __awaiter(this, void 0, void 0, function* () {
            const sceneFile = yield filesystem_1.readJSON(project_1.getSceneFilePath(this.workingDir));
            return this.validateParcelData(sceneFile);
        });
    }
    /**
     * Returns a random project name
     */
    getRandomName() {
        return dockerNames.getRandomName();
    }
    /**
     * Writes the `.dclignore` file to the provided directory path.
     * @param dir The target path where the file will be
     */
    writeDclIgnore() {
        return __awaiter(this, void 0, void 0, function* () {
            const content = [
                '.*',
                'package.json',
                'package-lock.json',
                'yarn-lock.json',
                'build.json',
                'tsconfig.json',
                'tslint.json',
                'node_modules/',
                '*.ts',
                '*.tsx',
                'dist/'
            ].join('\n');
            yield fs.outputFile(path.join(this.workingDir, project_1.DCLIGNORE_FILE), content);
            return content;
        });
    }
    /**
     * Validates all the conditions required for the creation of a new project.
     * Throws if a project already exists or if the directory is not empty.
     */
    validateNewProject() {
        return __awaiter(this, void 0, void 0, function* () {
            if ((yield this.sceneFileExists()) || (yield this.decentralandFolderExists())) {
                errors_1.fail(errors_1.ErrorType.PROJECT_ERROR, 'Project already exists');
            }
        });
    }
    /**
     * Validates all the conditions required to operate over an existing project.
     * Throws if a project contains an invalid main path or if the `scene.json` file is missing.
     */
    validateExistingProject() {
        return __awaiter(this, void 0, void 0, function* () {
            let sceneFile;
            try {
                sceneFile = yield filesystem_1.readJSON(project_1.getSceneFilePath(this.workingDir));
            }
            catch (e) {
                errors_1.fail(errors_1.ErrorType.PROJECT_ERROR, `Unable to read 'scene.json' file. Try initializing the project using 'dcl init'`);
            }
            if (!this.isWebSocket(sceneFile.main)) {
                if (!this.isValidMainFormat(sceneFile.main)) {
                    errors_1.fail(errors_1.ErrorType.PROJECT_ERROR, `Main scene format file (${sceneFile.main}) is not a supported format`);
                }
                if (!(yield this.fileExists(sceneFile.main))) {
                    errors_1.fail(errors_1.ErrorType.PROJECT_ERROR, `Main scene file ${sceneFile.main} is missing`);
                }
            }
        });
    }
    /**
     * Returns a promise of an array containing all the file paths for the given directory.
     * @param dir The given directory where to list the file paths.
     */
    getAllFilePaths(dir = this.workingDir) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const files = yield fs.readdir(dir);
                let tmpFiles = [];
                for (let i = 0; i < files.length; i++) {
                    const file = files[i];
                    const filePath = path.resolve(dir, file);
                    const relativePath = path.relative(this.workingDir, filePath);
                    const stat = yield fs.stat(filePath);
                    if (stat.isDirectory()) {
                        const folderFiles = yield this.getAllFilePaths(filePath);
                        tmpFiles = tmpFiles.concat(folderFiles);
                    }
                    else {
                        tmpFiles.push(relativePath);
                    }
                }
                return tmpFiles;
            }
            catch (e) {
                return [];
            }
        });
    }
    /**
     * Returns a promise of an array of objects containing the path and the content for all the files in the project.
     * All the paths added to the `.dclignore` file will be excluded from the results.
     * Windows directory separators are replaced for POSIX separators.
     * @param ignoreFile The contents of the .dclignore file
     */
    getFiles(ignoreFile) {
        return __awaiter(this, void 0, void 0, function* () {
            const files = yield this.getAllFilePaths();
            const filteredFiles = ignore()
                .add(ignoreFile)
                .filter(files);
            let data = [];
            for (let i = 0; i < filteredFiles.length; i++) {
                const file = filteredFiles[i];
                const filePath = path.resolve(this.workingDir, file);
                const stat = yield fs.stat(filePath);
                if (stat.size > Project.MAX_FILE_SIZE) {
                    // MAX_FILE_SIZE is an arbitrary file size
                    errors_1.fail(errors_1.ErrorType.IPFS_ERROR, `Maximum file size exceeded: '${file}' is larger than ${Project.MAX_FILE_SIZE} bytes`);
                }
                const content = yield fs.readFile(filePath);
                data.push({ path: file.replace(/\\/g, '/'), content: Buffer.from(content), size: stat.size });
            }
            return data;
        });
    }
    /**
     * Returns the the contents of the `.dclignore` file
     */
    getDCLIgnore() {
        return __awaiter(this, void 0, void 0, function* () {
            let ignoreFile;
            try {
                ignoreFile = yield fs.readFile(project_1.getIgnoreFilePath(this.workingDir), 'utf8');
            }
            catch (e) {
                ignoreFile = null;
            }
            return ignoreFile;
        });
    }
    /**
     * Returns `true` if the provided path contains a valid main file format.
     * @param path The path to the main file.
     */
    isValidMainFormat(path) {
        const supportedExtensions = new Set(['js', 'html', 'xml']);
        const mainExt = path ? path.split('.').pop() : null;
        return supportedExtensions.has(mainExt);
    }
    /**
     * Returns true if the given URL is a valid websocket URL.
     * @param url The given URL.
     */
    isWebSocket(url) {
        return /wss?\:\/\//gi.test(url);
    }
    /**
     * Returns `true` if the path exists as a valid file or websocket URL.
     * @param filePath The path to a given file.
     */
    fileExists(filePath) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.isWebSocket(filePath)) {
                return true;
            }
            return fs.pathExists(path.join(this.workingDir, filePath));
        });
    }
    /**
     * Fails the execution if one of the parcel data is invalid
     * @param sceneFile The JSON parsed file of scene.json
     */
    validateParcelData(sceneFile) {
        const { base, parcels } = sceneFile.scene;
        const parcelSet = new Set(parcels);
        if (!base) {
            errors_1.fail(errors_1.ErrorType.PROJECT_ERROR, 'Missing scene base attribute at scene.json');
        }
        if (!parcels) {
            errors_1.fail(errors_1.ErrorType.PROJECT_ERROR, 'Missing scene parcels attribute at scene.json');
        }
        if ([...parcelSet].length < parcels.length) {
            errors_1.fail(errors_1.ErrorType.PROJECT_ERROR, 'There are duplicated parcels at scene.json');
        }
        if (!parcelSet.has(base)) {
            errors_1.fail(errors_1.ErrorType.PROJECT_ERROR, `Your base parcel ${base} should be included on parcels attribute at scene.json`);
        }
        const objParcels = parcels.map(coordinateHelpers_1.getObject);
        objParcels.forEach(({ x, y }) => {
            if (coordinateHelpers_1.inBounds(x, y)) {
                return;
            }
            const { minX, maxX } = coordinateHelpers_1.getBounds();
            errors_1.fail(errors_1.ErrorType.PROJECT_ERROR, `Coordinates ${x},${y} are outside of allowed limits (from ${minX} to ${maxX})`);
        });
        if (!coordinateHelpers_1.areConnected(objParcels)) {
            errors_1.fail(errors_1.ErrorType.PROJECT_ERROR, 'Parcels described on scene.json are not connected. They should be one next to each other');
        }
    }
}
Project.MAX_FILE_SIZE = 524300000;
exports.Project = Project;
