// eth-connect/dist/Contract.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const SolidityFunction_1 = require("./SolidityFunction");
const SolidityEvent_1 = require("./SolidityEvent");
const AllSolidityEvents_1 = require("./AllSolidityEvents");
/**
 * Should be called to add functions to contract object
 *
 * @method addFunctionsToContract
 * @param {Contract} contract
 * @param {Array} abi
 */
function addFunctionsToContract(contract) {
    contract.abi
        .filter(function (json) {
        return json.type === 'function';
    })
        .map(function (json) {
        return new SolidityFunction_1.SolidityFunction(contract.requestManager, json, contract.address);
    })
        .forEach(function (f) {
        f.attachToContract(contract);
    });
}
/**
 * Should be called to add events to contract object
 *
 * @method addEventsToContract
 * @param {Contract} contract
 * @param {Array} abi
 */
function addEventsToContract(contract) {
    let events = contract.abi.filter(function (json) {
        return json.type === 'event';
    });
    let allEvents = new AllSolidityEvents_1.AllSolidityEvents(contract.requestManager, events, contract.address);
    allEvents.attachToContract(contract);
    events
        .map(function (json) {
        return new SolidityEvent_1.SolidityEvent(contract.requestManager, json, contract.address);
    })
        .forEach(function (e) {
        e.attachToContract(contract);
    });
}
/**
 * Should be called to create new contract instance
 *
 * @method Contract
 * @param {Array} abi
 * @param {Address} contract address
 */
class Contract {
    constructor(requestManager, abi, address) {
        this.requestManager = requestManager;
        this.abi = abi;
        this.address = address;
        this.events = {};
        this.transactionHash = null;
        this.transactionHash = null;
        this.address = address;
        this.abi = abi;
        // this functions are not part of prototype,
        // because we dont want to spoil the interface
        addFunctionsToContract(this);
        addEventsToContract(this);
    }
}
exports.Contract = Contract;
