// eth-connect/dist/index.js
"use strict";
/*
    This file is part of web3.js.

    web3.js is free software: you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    web3.js is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with web3.js.  If not, see <http://www.gnu.org/licenses/>.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const HTTPProvider_1 = require("./providers/HTTPProvider");
const WebSocketProvider_1 = require("./providers/WebSocketProvider");
const utils = require("./utils/utils");
exports.utils = utils;
var RequestManager_1 = require("./RequestManager");
exports.RequestManager = RequestManager_1.RequestManager;
var Contract_1 = require("./Contract");
exports.Contract = Contract_1.Contract;
var ContractFactory_1 = require("./ContractFactory");
exports.ContractFactory = ContractFactory_1.ContractFactory;
var Method_1 = require("./Method");
exports.Method = Method_1.Method;
var Property_1 = require("./Property");
exports.Property = Property_1.Property;
var eth_1 = require("./methods/eth");
exports.eth = eth_1.eth;
const providers = {
    HTTPProvider: HTTPProvider_1.HTTPProvider,
    WebSocketProvider: WebSocketProvider_1.WebSocketProvider
};
exports.providers = providers;
