// eth-connect/dist/Property.js
"use strict";
/*
    This file is part of web3.js.

    web3.js is free software: you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    web3.js is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with web3.js.  If not, see <http://www.gnu.org/licenses/>.
*/
Object.defineProperty(exports, "__esModule", { value: true });
class Property {
    constructor(options) {
        this.getter = options.getter;
        this.outputFormatter = options.outputFormatter || null;
    }
    /**
     * Should be called to format output(result) of method
     *
     * @method formatOutput
     * @param {object}
     * @return {object}
     */
    formatOutput(result) {
        return this.outputFormatter && result !== null && result !== undefined ? this.outputFormatter(result) : result;
    }
    async execute(requestManager) {
        const result = await requestManager.sendAsync({
            method: this.getter,
            params: []
        });
        return this.formatOutput(result);
    }
}
exports.Property = Property;
